plugins/kotlin/fir/src/org/jetbrains/kotlin/idea/completion/contributors/FirSuperMemberCompletionContributor.kt

```diff
@@ -6,54 +6,46 @@ import com.intellij.psi.util.parentsOfType
 import org.jetbrains.kotlin.analysis.api.KtAnalysisSession
 import org.jetbrains.kotlin.analysis.api.symbols.KtCallableSymbol
 import org.jetbrains.kotlin.analysis.api.symbols.KtFunctionSymbol
-import org.jetbrains.kotlin.analysis.api.types.KtType
 import org.jetbrains.kotlin.idea.completion.ItemPriority
 import org.jetbrains.kotlin.idea.completion.checkers.CompletionVisibilityChecker
 import org.jetbrains.kotlin.idea.completion.context.FirBasicCompletionContext
 import org.jetbrains.kotlin.idea.completion.context.FirNameReferencePositionContext
-import org.jetbrains.kotlin.idea.completion.contributors.helpers.collectNonExtensions
-import org.jetbrains.kotlin.idea.completion.lookups.CallableInsertionOptions
 import org.jetbrains.kotlin.idea.completion.lookups.CallableInsertionStrategy
-import org.jetbrains.kotlin.idea.completion.lookups.detectImportStrategy
 import org.jetbrains.kotlin.psi.KtNamedFunction
 import org.jetbrains.kotlin.psi.KtSuperExpression
 
 internal class FirSuperMemberCompletionContributor(
     basicContext: FirBasicCompletionContext,
     priority: Int
-) : FirCompletionContributorBase<FirNameReferencePositionContext>(basicContext, priority) {
+) : FirCallableCompletionContributor(basicContext, priority) {
+
+    /** A map that contains all containing functions as values, each of which is indexed by symbols it overrides. For example, consider
+     * the following code
+     * ```
+     * class A : Runnable {
+     *   override fun run() {
+     *     val o = object: Callable<String> {
+     *       override fun call(): String {
+     *         super.<caret>
+     *       }
+     *     }
+     *   }
+     * }
+     * ```
+     * The map would contain
+     *
+     * * Runnable.run -> A.run
+     * * Callable.call -> <anonymous object>.call
+     */
+    lateinit var superFunctionToContainingFunction: Map<KtCallableSymbol, KtNamedFunction>
```

Comment:
All completion contributors should be stateless. Why do you need to extract that local variable to a class member?

Reply:
I need to pass some additional data (`superFunctionToContainingFunction`) to `getInsertionStrategy`. The contributor is still stateless because `superFunctionToContainingFunction` is always updated in `complete` before used in `getInsertionStrategy`. Let me try refactoring this code to move all super related completion to this class instead. Then I probably won't need juggle all these data.